### PR TITLE
[Merged by Bors] - feat: port Data.Complex.Module

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Complex/Module.lean [hunk 1]
```diff
@@ -74,56 +74,56 @@ theorem real_smul {x : ℝ} {z : ℂ} : x • z = x * z :=
...
 instance [Semiring R] [DistribMulAction R ℝ] : DistribMulAction R ℂ :=
-  { Complex.distribSmul with }
+  { Complex.distribSMul, Complex.mulAction with }
 
 instance [Semiring R] [Module R ℝ] : Module R ℂ where
   add_smul r s x := by ext <;> simp [smul_re, smul_im, add_smul]
   zero_smul r := by ext <;> simp [smul_re, smul_im, zero_smul]
 
-instance [CommSemiring R] [Algebra R ℝ] : Algebra R ℂ :=
+noncomputable instance [CommSemiring R] [Algebra R ℝ] : Algebra R ℂ :=
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:101]: Any idea why this changed? This should get a porting note if nothing else.

File: Mathlib/Data/Complex/Module.lean [hunk 2]
```diff
@@ -171,28 +172,31 @@ theorem coe_basisOneI : ⇑basisOneI = ![1, I] :=
     Basis.apply_eq_iff.mpr <|
       Finsupp.ext fun j => by
         fin_cases i <;> fin_cases j <;>
-          simp only [coe_basis_one_I_repr, Finsupp.single_eq_of_ne, Matrix.cons_val_zero,
+          -- Porting note: consider squeezing again
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:175]:
```suggestion
          -- Porting note: removed `only`, consider squeezing again
```

File: Mathlib/Data/Complex/Module.lean [hunk 3]
```diff
@@ -171,28 +172,31 @@ theorem coe_basisOneI : ⇑basisOneI = ![1, I] :=
...
 #align complex.finrank_real_complex Complex.finrank_real_complex
 
 @[simp]
 theorem rank_real_complex : Module.rank ℝ ℂ = 2 := by simp [← finrank_eq_rank, finrank_real_complex]
 #align complex.rank_real_complex Complex.rank_real_complex
 
 theorem rank_real_complex'.{u} : Cardinal.lift.{u} (Module.rank ℝ ℂ) = 2 := by
-  simp [← finrank_eq_rank, finrank_real_complex, bit0]
+  rw [← finrank_eq_rank, finrank_real_complex, Cardinal.lift_natCast]
+  rfl
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean]: Is there a missing `simp` lemma here?

File: Mathlib/Data/Complex/Module.lean [hunk 4]
```diff
@@ -202,7 +206,8 @@ end Complex
 
 /- Register as an instance (with low priority) the fact that a complex vector space is also a real
 vector space. -/
-instance (priority := 900) Module.complexToReal (E : Type _) [AddCommGroup E] [Module ℂ E] :
+noncomputable instance (priority := 900) Module.complexToReal
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:209]: `noncomputable` here also worth a porting note: I assume it's the same reason.

Note that Lean4 no longer flags unecessary `noncomputable`s as far as I'm aware, hence the need for notes.

File: Mathlib/Data/Complex/Module.lean [hunk 5]
```diff
@@ -346,11 +354,11 @@ def liftAux (I' : A) (hf : I' * I' = -1) : ℂ →ₐ[ℝ] A :=
       rw [add_mul, mul_add, mul_add, add_comm _ (y₁ • I' * y₂ • I'), add_add_add_comm]
       congr 1
       -- equate "real" and "imaginary" parts
-      ·
-        rw [smul_mul_smul, hf, smul_neg, ← Algebra.algebraMap_eq_smul_one, ← sub_eq_add_neg, ←
-          RingHom.map_mul, ← RingHom.map_sub]
-      ·
-        rw [Algebra.smul_def, Algebra.smul_def, Algebra.smul_def, ← Algebra.right_comm _ x₂, ←
+      · -- Porting note: Replaced `smul_mul_smul` with
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:357]: Does `smul_mul_smul` behave with `etaExperiment`?

File: Mathlib/Data/Complex/Module.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/Complex/Module.lean]: Ah, I fixed this.

File: Mathlib/Data/Complex/Module.lean (hunk 5 above)
Comment (eric-wieser) [reply Mathlib/Data/Complex/Module.lean:355]: Adding this shortcut instance seems to help:
```
instance {R A} [CommSemiring R] [Semiring A] [Algebra R A] : SMulCommClass R A A :=
  by show_term infer_instance
```

File: Mathlib/Data/Complex/Module.lean [hunk 6]
```diff
@@ -337,6 +337,10 @@ section lift
 
 variable {A : Type _} [Ring A] [Algebra ℝ A]
 
+-- Porting note: added for `liftAux` proof
+local instance {R A} [CommSemiring R] [Semiring A] [Algebra R A] : SMulCommClass R A A := by
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:341]: I think we should backport this and add it globally. It's a prop-only instance and the most common use of smul_comm_clasa.

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Complex/Module.lean:341]: Like this? https://github.com/leanprover-community/mathlib/pull/18907

**Review by jcommelin — COMMENTED** [review]
The backport is in progress, any little tweaks there can be figured out independently of whether this is merged.

Thanks!

bors merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**